src/util/ser.rs

```diff
@@ -0,0 +1,373 @@
+impl_array!(32);
+impl_array!(33);
+
+// Tuples
+macro_rules! tuple_encode {
+	($($x:ident),*) => (
+		impl<W: Write, $($x: Writeable<W>),*> Writeable<W> for ($($x),*) {
+			#[inline]
+			#[allow(non_snake_case)]
+			fn write(&self, w: &mut Writer<W>) -> Result<(), Error> {
+					let &($(ref $x),*) = self;
+					$( $x.write(w)?; )*
+					Ok(())
+			}
+		}
+		impl<R: Read, $($x: Readable<R>),*> Readable<R> for ($($x),*) {
+			#[inline]
+			#[allow(non_snake_case)]
+			fn read(r: &mut Reader<R>) -> Result<($($x),*), Error> {
+					Ok(($({let $x = $x::read(r)?; $x }),*))
+			}
+		}
+	);
+}
+
+tuple_encode!(T0, T1);
+tuple_encode!(T0, T1, T2, T3);
+tuple_encode!(T0, T1, T2, T3, T4, T5);
+tuple_encode!(T0, T1, T2, T3, T4, T5, T6, T7);
+
+
+// References
+impl<W: Write, T: Writeable<W>> Writeable<W> for Box<T> {
+	#[inline]
+	fn write(&self, w: &mut Writer<W>) -> Result<(), Error> { (**self).write(w) }
+}
+
+impl<R: Read, T: Readable<R>> Readable<R> for Box<T> {
+	#[inline]
+	fn read(r: &mut Reader<R>) -> Result<Box<T>, Error> {
+		Ok(Box::new(<T as Readable<R>>::read(r)?))
+	}
+}
+
+// HashMap
+impl<W, K, V> Writeable<W> for HashMap<K, V>
+    where W: Write,
+          K: Writeable<W> + Eq + Hash,
+          V: Writeable<W>
+{
+    #[inline]
+    fn write(&self, w: &mut Writer<W>) -> Result<(), Error> {
+				(self.len() as u16).write(w)?;
+        for (key, value) in self.iter() {
+            key.write(w)?;
+            value.write(w)?;
+        }
+        Ok(())
+    }
+}
+
+impl<R, K, V> Readable<R> for HashMap<K, V>
+    where R: Read,
+          K: Readable<R> + Eq + Hash,
+          V: Readable<R>
+{
+	#[inline]
+	fn read(r: &mut Reader<R>) -> Result<HashMap<K, V>, Error> {
+		let len: u16 = Readable::read(r)?;
+		let mut ret = HashMap::with_capacity(len as usize);
+		for _ in 0..len {
+				ret.insert(K::read(r)?, V::read(r)?);
+		}
+		Ok(ret)
+	}
+}
+
+impl<W: Write, T: Writeable<W>> Writeable<W> for [T] {
+	#[inline]
+	fn write(&self, w: &mut Writer<W>) -> Result<(), Error> {
+		(self.len() as u16).write(w)?;
+		for c in self.iter() { c.write(w)?; }
+		Ok(())
+	}
+}
+
+// Vectors
+impl<W: Write, T: Writeable<W>> Writeable<W> for Vec<T> {
+	#[inline]
+  fn write(&self, w: &mut Writer<W>) -> Result<(), Error> { 
+		(&self[..]).write(w)
+	}
+}
+
+impl<R: Read, T: Readable<R>> Readable<R> for Vec<T> {
+	#[inline]
+	fn read(r: &mut Reader<R>) -> Result<Vec<T>, Error> {
+			let len: u16 = Readable::read(r)?;
+			let byte_size = (len as usize)
+												.checked_mul(mem::size_of::<T>())
+												.ok_or(Error::InvalidLength)?;
+			if byte_size > MAX_VEC_SIZE {
+					return Err(Error::InvalidLength);
+			}
+			let mut ret = Vec::with_capacity(len as usize);
+			for _ in 0..len { ret.push(T::read(r)?); }
+			Ok(ret)
+	}
+}
+
+impl<W: Write, T: Writeable<W>> Writeable<W> for Option<T> {
+    #[inline]
+    fn write(&self, w: &mut Writer<W>) -> Result<(), Error> {
+        match self {
+            Some(data) => {
+                true.write(w)?;
```

Comment:
I dont think any of the Option<>s in msgs have a flag byte. Instead they simply deserialize if there are more bytes to be read, and None if there are not.

Reply:
Removed `Option<?>` altogether as it seems difficult to me implementing generic serialization varying depending on the type of the payload of `Option`. But I left something equivalent to `Option<Script>` which found usage in two messages.